drafts/str4d-overwinter-activation/draft1.rst

```diff
@@ -0,0 +1,229 @@
+  the first software version containing support for the network upgrade is released, for the following reason:
+
+  - As of the time of writing (the 1.0.15 release), the release cycle is six weeks long, and nodes undergo
+    auto-senescence 16 weeks after release. Thus, if version ``X`` contains support for a network upgrade,
+    version ``X-1`` will be deprecated 10 weeks after the release of version ``X``, which is about 2.3 months.
+    A three-month window provides ample time for users to upgrade their nodes after auto-senescence, and
+    re-integrate into the network prior to activation of the network upgrade.
+
+The relationship between ``BRANCH_ID`` and ``ACTIVATION_HEIGHT`` is many-to-one: it is possible for many
+distinct branches to descend from the same parent block (and thus have the same ``ACTIVATION_HEIGHT``), but a
+specific branch can only have one parent block. Concretely, this means that if the ``ACTIVATION_HEIGHT`` of a
+network upgrade is changed for any reason (e.g. security vulnerabilities or consensus bugs are discovered),
+the ``BRANCH_ID`` MUST also be changed.
+
+Activation mechanism
+--------------------
+
+A blockchain is defined as invalid if, within the set of all network upgrades that have activated in the past
+(or will activate in future) on that chain, an ``ACTIVATION_HEIGHT`` is repeated. Note that this does not
+require ``ACTIVATION_HEIGHT`` to be globally unique, or even locally unique; multiple network upgrades can
+occur in parallel, as long as they are non-overlapping (only one will activate on any given chain).
+
+Concretely, this means that the Zcash blockchain is broken into "epochs" of block height intervals
+``[ACTIVATION_HEIGHT_N, ACTIVATION_HEIGHT_{N+1})`` (i.e. including ``ACTIVATION_HEIGHT_N`` and excluding
+``ACTIVATION_HEIGHT_{N+1}``), on which consensus rule sets are defined.
+
+Consensus rules themselves (and any network behavior or surrounding code that depends on them) MUST be gated
+by block height checks. For example:
+
+.. code:: cpp
+
+  if (CurrentEpoch(chainActive.Height(), Params().GetConsensus()) == Consensus::UPGRADE_OVERWINTER) {
+      // Overwinter-specific logic
+  } else {
+      // Non-Overwinter logic
+  }
+
+  // ...
+
+  if (NetworkUpgradeActive(pindex->nHeight, Params().GetConsensus(), Consensus::UPGRADE_OVERWINTER)) {
+      // Overwinter consensus rules applied to block
+  } else {
+      // Pre-Overwinter consensus rules applied to block
+  }
+
+
+Block validation
+````````````````
+Incoming blocks known to have a particular height (due to their parent chain being entirely known) MUST be
+validated under the consensus rules corresponding to the expected branch ID for that height.
+
+Incoming blocks with unknown heights (because at least one block header in their parent chain is unknown)
+MAY be cached for future consideration after all their parents have been received.
```

Comment:
There's an embiguity here:
* "MAY be (cached for future consideration) after all their parents have been received" (not intended) vs
* "MAY be cached for (consideration in the future after all their parents have been received)" (intended).

Change to "MAY be cached, so that they can be reconsidered in the future after all their parents have been received".